src/main/java/ch/njol/skript/hooks/permission/expressions/ExprGroup.java

```diff
@@ -38,54 +45,102 @@
 @Description("The primary group or all groups of a player. This expression requires Vault and a compatible permissions plugin to be installed.")
 @Examples({"on join:",
-            "broadcast \"%group of player%\" # this is the player's primary group",
-            "broadcast \"%groups of player%\" # this is all of the player's groups"})
+			"broadcast \"%group of player%\" # this is the player's primary group",
+			"broadcast \"%groups of player%\" # this is all of the player's groups"})
 @Since("INSERT VERSION")
-public class ExprGroup extends PropertyExpression<Player, String> {
+public class ExprGroup extends SimpleExpression<String> {
 
-    static {
-        register(ExprGroup.class, String.class, "group[(1¦s)]", "players");
-    }
+	static {
+		PropertyExpression.register(ExprGroup.class, String.class, "group[(1¦s)]", "offlineplayers");
+	}
 
-    private boolean primary;
+	private boolean primary;
+	@Nullable
+	private Expression<OfflinePlayer> players;
 
-    @SuppressWarnings("null")
-    @Override
-    public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
-        setExpr((Expression<Player>) exprs[0]);
-        primary = parseResult.mark == 0;
-        return true;
-    }
+	@SuppressWarnings({"null", "unchecked"})
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
+		if (!VaultHook.permission.hasGroupSupport()) {
+			Skript.error(VaultHook.NO_GROUP_SUPPORT);
+			return false;
+		}
+		players = (Expression<OfflinePlayer>) exprs[0];
+		primary = parseResult.mark == 0;
+		return true;
+	}
 
-    @SuppressWarnings("null")
-    @Override
-    protected String[] get(Event e, Player[] players) {
-        List<String> groups = new ArrayList<>();
-        for (Player player : players) {
-            if (primary)
-                groups.add(VaultHook.permission.getPrimaryGroup(player));
-            else
-                Collections.addAll(groups, VaultHook.permission.getPlayerGroups(player));
-        }
-        return groups.toArray(new String[groups.size()]);
-    }
+	@SuppressWarnings("null")
+	@Override
+	protected String[] get(Event e) {
+		List<String> groups = new ArrayList<>();
+		for (OfflinePlayer player : players.getArray(e)) {
+			if (primary)
+				groups.add(VaultHook.permission.getPrimaryGroup(null, player));
+			else
+				Collections.addAll(groups, VaultHook.permission.getPlayerGroups(null, player));
+		}
+		return groups.toArray(new String[0]);
+	}
 
-    @SuppressWarnings("null")
-    @Override
-    public boolean isSingle() {
-        return super.isSingle() && primary;
-    }
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(Changer.ChangeMode mode) {
+		if (mode == Changer.ChangeMode.ADD ||
+				mode == Changer.ChangeMode.REMOVE ||
+				mode == Changer.ChangeMode.SET ||
+				mode == Changer.ChangeMode.DELETE ||
+				mode == Changer.ChangeMode.RESET) {
```

Comment:
Why not just `if (mode != Changer.ChangeMode.REMOVE_ALL)` ?

Reply:
~~what if new change modes are added~~

Reply:
Then half of Skript breaks... This being the easiest thing to fix.